### PR TITLE
Message router for LCMP & XCMP

### PR DESCRIPTION
**Pull request by jiguantong**
https://github.com/darwinia-network/darwinia/issues/1090

TODO
- [x] New barrier for XCM `DescendOrigin` instruction
- [x] Support AccountKey20 > AccountId32
- [x] Route SmartChain > Moonbeam
- [x] Benchmark & update weight
- [x] Support pangolin <> pangolin parachain alpha bridge
       https://github.com/darwinia-network/darwinia-common/pull/1453
       https://github.com/darwinia-network/darwinia-messages-substrate/pull/190
- [x] Test on Moonbase Alpha

### CODE REVIEW

File: runtime/common/Cargo.toml [hunk 1]
```diff
@@ -34,15 +34,20 @@ frame-system               = { default-features = false, git = "https://github.c
...
+polkadot-core-primitives    = { default-features = false, git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.27" }
+polkadot-parachain          = { default-features = false, git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.27" }
+polkadot-runtime-parachains = { default-features = false, git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.27" }
+xcm                         = { default-features = false, git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.27" }
+xcm-builder                 = { default-features = false, git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.27" }
+xcm-executor                = { default-features = false, git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.27" }
 
 [dev-dependencies]
 pallet-timestamp = { default-features = false, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.27" }
-sp-io            = { default-features = false, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.27" }
```
Comment (xiaoch05) [inline runtime/common/Cargo.toml]: this will be used in parachain issuing pallet

File: runtime/common/src/message_router/mod.rs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+			Success = MultiLocation,
+		>;
+		type LocalAssetId: Get<MultiLocation>;
+		/// Used to calculate the weight required for the local execution of xcm.
+		type LocalWeigher: WeightBounds<Self::Call>;
+		/// Means of inverting a location.
+		type LocationInverter: InvertLocation;
+		type MoonbeamLocation: Get<MultiLocation>;
+		/// Used to calculate the weight required for the moonbeam execution of xcm.
+		type MoonbeamWeigher: WeightBounds<Self::Call>;
```
Comment (xiaoch05) [inline runtime/common/src/message_router/mod.rs:66]: Can we do generalization here?

File: runtime/common/src/message_router/mod.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+					local_asset_units_per_second =
+						TargetXcmExecConfig::<T>::get(T::MoonbeamLocation::get())
+							.ok_or(Error::<T>::TargetXcmExecNotConfig)?;
+					remote_weight = T::MoonbeamWeigher::weight(&mut Self::extend_remote_xcm(
+						raw_account.clone(),
+						remote_xcm.clone(),
+						MultiAsset { id: AssetId::from(T::LocalAssetId::get()), fun: Fungible(0) },
+					))
+					.map_err(|()| Error::<T>::UnweighableMessage)?;
+				},
```
Comment (xiaoch05) [inline runtime/common/src/message_router/mod.rs:182]: What about the processing of other branches

File: runtime/common/Cargo.toml (hunk 1 above)
Comment (jiguantong) [reply runtime/common/Cargo.toml]: ok I will add it back

File: runtime/common/src/message_router/location_conversion.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+use xcm_executor::traits::Convert;
+
+/// The address prefix for dvm address
+const ADDR_PREFIX: &[u8] = b"dvm:";
+
+/// Derive AccountKey20 in MultiLocation to AccountId32
+/// Refer to https://github.com/darwinia-network/darwinia-common/blob/main/frame/support/src/evm.rs#L85
+pub struct AccountKey20Derive<AccountId>(PhantomData<AccountId>);
+impl<AccountId: From<[u8; 32]> + Into<[u8; 32]> + Clone> Convert<MultiLocation, AccountId>
+	for AccountKey20Derive<AccountId>
```
Comment (hackfisher) [inline runtime/common/src/message_router/location_conversion.rs:30]: This is used for converting other key types to Darwinia's AccountId32 format and compatible with H160 converting at the same time right?

File: runtime/common/src/message_router/location_conversion.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		};
+		let mut raw_account = [0u8; 32];
+
+		raw_account[0..4].copy_from_slice(ADDR_PREFIX);
+		raw_account[11..31].copy_from_slice(&key[..]);
+		raw_account[31] = checksum_of(&raw_account);
+		Ok(raw_account.into())
+	}
+
+	fn reverse(who: AccountId) -> Result<MultiLocation, AccountId> {
```
Comment (hackfisher) [inline runtime/common/src/message_router/location_conversion.rs:50]: What is reverse then? Do we need to convert back to moonbeam's H160?

File: runtime/common/src/message_router/mod.rs [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+			})?;
+
+			// Toggle the xcm_fee relative to a target context
+			let ancestry = T::LocationInverter::ancestry();
+			let mut remote_xcm_fee_anchor_dest = remote_xcm_fee.clone();
+			match target {
+				Target::Moonbeam => {
+					remote_xcm_fee_anchor_dest
+						.reanchor(&T::MoonbeamLocation::get(), &ancestry)
+						.map_err(|()| Error::<T>::MultiLocationFull)?;
```
Comment (hackfisher) [inline runtime/common/src/message_router/mod.rs:237]: What is `reanchor`? Could you give more explain for this part?

**Review by hackfisher — COMMENTED** [review]
Having some questions, I'm still learning this PR and may take another some time for learning before merge.

@AurevoirXavier @wuminzhe Please also review this PR.

File: runtime/common/src/message_router/mod.rs (hunk 3 above)
Comment (jiguantong) [reply runtime/common/src/message_router/mod.rs:182]: no other branch

File: runtime/common/src/message_router/location_conversion.rs (hunk 4 above)
Comment (jiguantong) [reply runtime/common/src/message_router/location_conversion.rs:30]: This is just used for converting [u8;20] (compatible with H160) key type to AccountId32 format.

File: runtime/common/src/message_router/location_conversion.rs (hunk 5 above)
Comment (jiguantong) [reply runtime/common/src/message_router/location_conversion.rs:50]: We dont need it.

**Review by hackfisher — APPROVED** [review]

**Pull request merged**